### PR TITLE
[1.1] Make reshape_op reuse input.

### PR DESCRIPTION
**Pull request by guoshengCS**
Make reshape_op reuse input to avoid memory copying, which speeds up Transformer significantly.

### CODE REVIEW

File: python/paddle/fluid/layers/nn.py [hunk 1]
```diff
@@ -4892,15 +4892,22 @@ def reshape(x, shape, actual_shape=None, act=None, inplace=True, name=None):
...
-                       a new output tensor is created
-                       whose data is copied from input x.
+        act (str): The non-linear activation to be applied to the reshaped tensor
+                   variable.
+        inplace(bool): If this flag is set true, reuse input :attr:`x` to reshape,
+                       which will change the shape of tensor variable :attr:`x`.
+                       Otherwise, preserve the shape :attr:`x` and create a new
+                       output tensor variable whose data is copied from input x
+                       but reshaped. Though setting to :attr:`True` will be more
+                       efficient, :attr:`False` is suggested when :attr:`x` are
```
Comment (panyx0718) [inline python/paddle/fluid/layers/nn.py]: what does `suggested` mean? Does it mean the result can be wrong when `x` is used in multiple operators? If true, you should be very clear in the doc.

File: python/paddle/fluid/layers/nn.py [hunk 2]
```diff
@@ -4844,7 +4844,7 @@ def autoincreased_step_counter(counter_name=None, begin=1, step=1):
     return counter
 
 
-def reshape(x, shape, actual_shape=None, act=None, inplace=True, name=None):
```
Comment (panyx0718) [inline python/paddle/fluid/layers/nn.py:4847]: Does the `true` do nothing in the past?

File: python/paddle/fluid/layers/nn.py [hunk 3]
```diff
@@ -4844,7 +4844,7 @@ def autoincreased_step_counter(counter_name=None, begin=1, step=1):
...
+def reshape(x, shape, actual_shape=None, act=None, inplace=False, name=None):
```
Comment (panyx0718) [inline python/paddle/fluid/layers/nn.py:4847]: Another question. should `inplace` be decided by memory optimizer? So, it's not decided by the API

File: python/paddle/fluid/layers/nn.py (hunk 2 above)
Comment (guoshengCS) [reply python/paddle/fluid/layers/nn.py:4847]: Yes, it is.

File: python/paddle/fluid/layers/nn.py (hunk 3 above)
Comment (guoshengCS) [reply python/paddle/fluid/layers/nn.py:4847]: Currently, the memory optimizer can't reuse the input variable directly. We will refine it later.

File: python/paddle/fluid/layers/nn.py (hunk 1 above)
Comment (guoshengCS) [reply python/paddle/fluid/layers/nn.py]: Done. Clarify the danger in the usage of `inplace`.

**Review by panyx0718 — APPROVED** [review]
LG since `inplace=True ` doesn't do anything in the past, won't break any user

**Review by qingqing01 — APPROVED** [review]

**Pull request merged**